### PR TITLE
Allow for specification of BASE_URL via container environment variable

### PR DESCRIPTION
**Pull request by mrsegen**
### Description

- Bring in Nginx configuration approach from swagger-api/swagger-ui repository.
- Update Dockerfile appropriately
- Add flags within index.html to allow configurator/index.js to attempt to replace text should users choose to. (I imagine this might address an issue, but I have not searched for such.)
- Modify configurator/translator.js to match the current values within index.html's window.onload

@fehguy Because you're listed as maintainer for the Dockerfile, I want to let you know about this PR. (To others, though, please note that this pattern used is identical to what is in [the analogous Dockerfile](https://github.com/swagger-api/swagger-ui/blob/master/Dockerfile) for which fehguy is also listed as maintainer.)

### Motivation and Context
This adds the option to specify a BASE_URL at which to serve the application. As is done with Swagger UI (see it's [Installation documentation](https://github.com/swagger-api/swagger-ui/blob/5b70866f7cd26e1eec449dd6d25677a11d753714/docs/usage/installation.md#docker)), the above PR now allows specification of a BASE_URL. This allows the editor to be served at a different location (e.g. `/swagger-editor`), which can enable this to be served at a single URL routed via a Kubernetes Ingress object.

Fix: #1956 

I had hoped to also address #1969 by simply copying in [that file](https://github.com/swagger-api/swagger-ui/blob/master/dist/oauth2-redirect.html) to this repository, but I am not yet prepared to test that responsibly. It is mentioned in the js/js.map bundles.)
This may or may not also help with #1510. I hope that it does, but that's also outside the scope of current focus.

### How Has This Been Tested?
1. `docker build .`
2. `docker run -e BASE_URL='/swagger-editor' -p 8001:8080 86547a8f42b1`
3. Navigate to http://localhost:8001/swagger-editor and see that the page loads

### My PR contains... 

- [x] No code changes (`src/` is unmodified: changes to documentation, CI, metadata, etc.)
- [ ] Dependency changes (any modification to dependencies in `package.json`)
- [ ] Bug fixes (non-breaking change which fixes an issue)
- [ ] Improvements (misc. changes to existing features)
- [x] Features (non-breaking change which adds functionality)

### My changes...
- [ ] are breaking changes to a public API (config options, System API, major UI change, etc).
- [ ] are breaking changes to a private API (Redux, component props, utility functions, etc.).
- [ ] are breaking changes to a developer API (npm script behavior changes, new dev system dependencies, etc).
- [x] are not breaking changes.

^ I don't _know_ of any breaking changes. Admittedly, there are non-trivial changes to the container image. Though, note that it's very similar to what is in place in Swagger Editor.

### Documentation
- [ ] My changes do not require a change to the project documentation.
- [x] My changes require a change to the project documentation.
- [x] If yes to above: I have updated the documentation accordingly.

### Automated tests
- [ ] My changes can not or do not need to be tested.
- [ ] My changes can and should be tested by unit and/or integration tests.
- [ ] If yes to above: I have added tests to cover my changes.
- [ ] If yes to above: I have taken care to cover edge cases in my tests.
- [ ] All new and existing tests passed.

^Up for consultation. I notice early pushes of this PR passed all checks even though the built image wasn't yet functional. If anyone who knows of approaches to test built docker images - that are super easy, please let me know. Otherwise, I'd ask that you please help more concretely.

### CODE REVIEW

**Comment by mrsegen** [comment]
Dockerfile seemingly doesn't work locally. I'm surprised tests pass in that case.

This may indicate an need for an additional workflow check in that area.

File: docker/configurator/variables.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
+const standardVariables = {
```
Comment (mrsegen) [inline docker/configurator/variables.js:1]: If there's a reason to reduce which of these are being added, please let me know.

File: docker/configurator/variables.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+  },
+  SHOW_COMMON_EXTENSIONS: {
+    type: "boolean",
+    name: "showCommonExtensions"
+  },
+  USE_UNSAFE_MARKDOWN: {
+    type: "boolean",
+    name: "useUnsafeMarkdown"
+  },
+  OAUTH2_REDIRECT_URL: {
```
Comment (mrsegen) [inline docker/configurator/variables.js:78]: Seems worth including insofar as it may help with #1969 and #1510

File: docker/nginx.conf [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  gzip_static on;
+  gzip_disable "msie6";
+
+  gzip_vary on;
+  gzip_types text/plain text/css application/javascript;
+
+  map $request_method $access_control_max_age {
+    OPTIONS 1728000; # 20 days
+  }
+  add_header X-Frame-Options deny;
```
Comment (mrsegen) [inline docker/nginx.conf:25]: I'm similarly unsure about whether this is still needed. It's carry over from the original docker-run.sh in this repository.

File: docker/nginx.conf [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+  gzip on;
+  gzip_static on;
+  gzip_disable "msie6";
+
+  gzip_vary on;
+  gzip_types text/plain text/css application/javascript;
+
+  map $request_method $access_control_max_age {
+    OPTIONS 1728000; # 20 days
+  }
```
Comment (mrsegen) [inline docker/nginx.conf:24]: I don't know the implications of this, but brought it in because it was expected by other parts of the "configurator" code brought in from swagger-ui

File: docker/run.sh [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+node /usr/share/nginx/configurator $INDEX_FILE
+
+# TODO: this is empty but we'll be adding configuration values here
+
+if [[ "${BASE_URL}" != "/" ]]; then
+  sed -i "s|location / {|location $BASE_URL {|g" $NGINX_CONF
+fi
+
+## Adding env var support for swagger file (json or yaml)
+if [[ -f "$SWAGGER_FILE" ]]; then
```
Comment (mrsegen) [inline docker/run.sh:18]: I have not yet tested whether this preserves or breaks this `docker run -d -p 80:8080 -v $(pwd):/tmp -e SWAGGER_FILE=/tmp/swagger.json swaggerapi/swagger-editor` functionality.

**Comment by mrsegen** [comment]
I acknowledge there is the tangentially related PR https://github.com/swagger-api/swagger-editor/pull/2073 .

I anticipate no conflict. In fact, before I was considering adding that change (adding the oauth2 redirect html  file) to this branch.

File: docker/nginx.conf (hunk 4 above)
Comment (mrsegen) [reply docker/nginx.conf:24]: *Part of the cors.conf file, not the configurator js files

File: docker/nginx.conf (hunk 3 above)
Comment (mrsegen) [reply docker/nginx.conf:25]: I don't see why that would need to be here. (It prevents the tool from being included in an iframe.) But any change to that functionality would be beyond the scope of this.

File: docker/configurator/variables.js (hunk 2 above)
Comment (mrsegen) [reply docker/configurator/variables.js:78]: This may be applicable to the functionality being added with https://github.com/swagger-api/swagger-editor/pull/2073

**Comment by mrsegen** [comment]
Closing this in favor of a cleaner PR: https://github.com/swagger-api/swagger-editor/pull/2281
